options-manager.js

```diff
@@ -230,20 +231,8 @@ const getGitIgnores = opts => globby
 		ignore: opts.ignores || [],
 		cwd: opts.cwd || process.cwd()
 	})
-	.map(pathToGitignore => {
-		const patterns = parseGitignore(pathToGitignore);
-		const base = path.dirname(pathToGitignore);
-
-		return patterns
-			.map(pattern => {
-				const negate = !pattern.startsWith('!');
-				const patternPath = negate ? pattern : pattern.substr(1);
-				return {negate, pattern: path.join(base, patternPath)};
-			})
-			.sort(pattern => pattern.negate ? 1 : -1)
-			.map(item => item.negate ? `!${item.pattern}` : item.pattern);
-	})
-	.reduce((a, b) => a.concat(b), []);
+	.map(new GitignoreParser(opts).parseFile)
```

Comment:
@marionebl The alternative of this line would have been:
```js
.map(filepath => gitignoreParserHelper.parseFile(filepath, opts)) // or similar
```

It is more a matter of taste, but this would result in unnecessary `opts = Object.assign({cwd: ''}, opts)` calls **for each `filepath`** instead of one.

So I decided to use [partial application](https://en.wikipedia.org/wiki/Partial_application) instead. I've had 2 choices:
1. use higher-order functions
2. use a `class` (which [is just syntactic sugar for `function`](https://babeljs.io/learn-es2015/#ecmascript-2015-features-classes); thus, to some extend, it's also a higher-order function)

Alternative 1. would look like this:

```js
const applyDefaults = opts => Object.assign({cwd: ''}, opts); // almost identically to `GitignoreParser.constructor`

exports.parsePattern = (pattern, baseDirectory) => {
	// identically to `GitignoreParser.parsePattern`
};

exports.parseFile = opts => {
	opts = applyDefaults(opts);

	return filepath => {
		// almost identically to `GitignoreParser.parseFile`
	};
};
```

It should be easy to see how the 1st and 2nd alternative correspond to each other.

I've chosen `class` because of:
1. IMO it is more readable.
2. `GitignoreParser` make sense to me – maybe just because I've already written some `XYZParsers` in less functional programming languages.
3. `class` improves the structures the code: the scope of `exports.parseFile` would be the whole file, `obj.parseFile` is only valid if `obj` is a `GitignoreParser` instance.
4. I had no better idea :wink:

@marionebl I do not feel strongly about `class` – feel free to share your thoughts.